### PR TITLE
feat: view presto row objects in data grid

### PR DESCRIPTION
**Pull request by khtruong**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

A row or array column is displayed in a single column. This was difficult to comprehend if there were many fields in the row or array.

**Example:**
Name: ColumnA
Data type: row(nested_column1 float, rowA row(rowB row(nested_column2 int)))

Previous Data Grid Display:
Screen Shot 2019-05-02 at 2 49 51 PM

This change will separate out row fields into its own column.

New Data Grid Display: 
Screen Shot 2019-05-02 at 2 49 36 PM

If the array is an array of row objects, we just display the whole thing like before. Ideally we would like to see that information more clearly but that is tracked by another story.

The changes look like a lot but really we are just doing the following 2 things:
1. Filter out any columns related to arrays of row objects. For example:
Name: ColumnB
Data type: array(row(nested_column int))
We have column objects for ColumnB and ColumnB.nested_column but we ignore the object corresponding to ColumnB.nested_column because we cannot easily create a select statement for such columns.
2. Create column clauses correctly where column names in dot notation are quoted and we provide a label for the column. For example:
ColumnA.nested_object would be represented as "ColumnA"."nested_object" AS "ColumnA.nested_object"

### TEST PLAN
Manually and added unit tests

### REVIEWERS
@betodealmeida @DiggidyDave @bearcage @xtinec @datability-io @john-bodley @michellethomas @graceguo-supercat @mistercrunch 

@williaster @kristw I made a small change to the data grid where headers are elided from the left if they are too long.

### CODE REVIEW

File: superset/assets/src/components/FilterableTable/FilterableTableStyles.css [hunk 1]
```diff
@@ -72,3 +72,10 @@
 }
 .even-row { background: #f2f2f2; }
 .odd-row { background: #ffffff; }
+.header-style {
+  direction: rtl;
+  overflow: hidden;
```
Comment (betodealmeida) [inline superset/assets/src/components/FilterableTable/FilterableTableStyles.css]: Should we do `overflow: scroll` here instead? I know it doesn't look nice, but at least it allows people to different between columns that have long names and similar prefixes.

File: superset/assets/src/components/FilterableTable/FilterableTableStyles.css [hunk 2]
```diff
@@ -72,3 +72,10 @@
 }
 .even-row { background: #f2f2f2; }
 .odd-row { background: #ffffff; }
+.header-style {
+  direction: rtl;
```
Comment (betodealmeida) [inline superset/assets/src/components/FilterableTable/FilterableTableStyles.css]: Do we need this? Superset supports different translations, and potentially some of them could be `rtl`.

File: superset/db_engine_specs.py [hunk 3]
```diff
@@ -980,19 +981,82 @@ def get_columns(
...
+
+    @classmethod
+    def _get_fields(cls, cols: List[dict]) -> List[ColumnClause]:
+        """
+        Format column clauses where names are in quotes and labels are specified
+        :param cols: columns
+        :return: column clauses
+        """
+        column_clauses = []
+        dot_regex = r'\.(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)'
```
Comment (betodealmeida) [inline superset/db_engine_specs.py]: Nit, Python supports verbose regular expressions: https://docs.python.org/3/library/re.html#re.VERBOSE

I think it would be good to transform this into a verbose regex, since it's really heard to understand what's going on here.

File: superset/db_engine_specs.py [hunk 4]
```diff
@@ -980,19 +981,82 @@ def get_columns(
...
+        for col in cols:
+            # get individual column names
+            col_names = re.split(dot_regex, col['name'])
+            # quote each column name if it is not already quoted
+            for index, col_name in enumerate(col_names):
+                if not cls._is_column_name_quoted(col_name):
+                    col_names[index] = '"{}"'.format(col_name)
```
Comment (betodealmeida) [inline superset/db_engine_specs.py]: Nit, you can use a list comprehension here (or at least an f-string):

```python
col_names = [
  col_name if cls._is_column_name_quoted(col_name) else f'"{col_name}"'
  for col_name in col_names
]
```

File: superset/db_engine_specs.py [hunk 5]
```diff
@@ -980,19 +981,82 @@ def get_columns(
...
+            quoted_col_name = '.'.join(col_names)
```
Comment (betodealmeida) [inline superset/db_engine_specs.py]: Nit, you can use a generator expression here (and also an f-string):

```python
quoted_col_name = '.'.join(
  col_name if cls._is_column_name_quoted(col_name) else f'"{col_name}"'
  for col_name in col_names
)
```

File: superset/db_engine_specs.py [hunk 6]
```diff
@@ -1526,6 +1590,10 @@ def where_latest_partition(
                     return qry.where(Column(col_name) == value)
         return False
 
+    @classmethod
+    def _get_fields(cls, cols: List[dict]) -> List[ColumnClause]:
+        return BaseEngineSpec._get_fields(cols)
+
```
Comment (betodealmeida) [inline superset/db_engine_specs.py]: Why is this needed? Isn't the class derived from `BaseEngineSpec`?

File: superset/db_engine_specs.py [hunk 7]
```diff
@@ -980,19 +981,82 @@ def get_columns(
...
+            column_clause = sqla.literal_column(quoted_col_name).label(col['name'])
+            column_clauses.append(column_clause)
+        return column_clauses
+
+    @classmethod
+    def _filter_presto_cols(cls, cols: List[dict]) -> List[dict]:
+        """
+        We want to filter out columns that correspond to array content because you cannot
+        select array content without an index, which will lead to a large and complicated
+        query. We know which columns to skip because cols is a list provided to us in a
```
Comment (betodealmeida) [inline superset/db_engine_specs.py]: @khtruong, can you explain what you mean with "you cannot select array content without an index, which will lead to a large and complicated query."?

**Comment by betodealmeida** [comment]
@khtruong, this LGTM but I left a few minor comments.

File: superset/db_engine_specs.py (hunk 6 above)
Comment (khtruong) [reply superset/db_engine_specs.py]: The class is derived from PrestoEngineSpec, which is then derived from BaseEngineSpec.

File: superset/db_engine_specs.py (hunk 7 above)
Comment (khtruong) [reply superset/db_engine_specs.py]: Lol, yeah I'm not surprised that it is still confusing. Let me re-phrase it.

File: superset/db_engine_specs.py (hunk 5 above)
Comment (khtruong) [reply superset/db_engine_specs.py]: Sweet. Thanks. My python is a bit rusty so this is definitely more pythonic.

File: superset/assets/src/components/FilterableTable/FilterableTableStyles.css (hunk 1 above)
Comment (khtruong) [reply superset/assets/src/components/FilterableTable/FilterableTableStyles.css]: I tried adding that style but it looked extremely bad :(

File: superset/assets/src/components/FilterableTable/FilterableTableStyles.css (hunk 2 above)
Comment (khtruong) [reply superset/assets/src/components/FilterableTable/FilterableTableStyles.css]: Good point. Maybe I should just make it the standard ellipsis for now to be consistent with the rest of the app because the answer to your question should be implemented throughout the product.

File: superset/db_engine_specs.py (hunk 6 above)
Comment (betodealmeida) [reply superset/db_engine_specs.py]: Ah, got it! :)

**Review by betodealmeida — APPROVED** [review]
Thanks, @khtruong
